### PR TITLE
Added always_uses_dpop client registration metadata parameter

### LINKED ISSUE #105: client metadata
This was kind of mentioned in #27 but I think not taken forward as use cases couldn't be seen.

I'm not sure if I've missed something but it seems like there's an argument that if a client supports dpop access tokens, it should be able to indicate so to the server, with an expectation that (whether as a result of a mistake by the client developer, a misconfiguration of the client, or some kind of attack*) that the server could decide not to issue a plain bearer access token.

So similar to  [tls_client_certificate_bound_access_tokens](https://www.rfc-editor.org/rfc/rfc8705.html#name-dynamic-client-registration) I wonder about a `dpop_access_tokens` or similar. But perhaps more generic as there should probably be a way to indicate the use of dpop refresh tokens for public clients too.

[*] admittedly I can't see any mechanism for making such an attack, it seems like an attacker that could force the client to request a bearer token would also be in a position to inject their own public key - perhaps if the server had some kind of requirement that key had been attested to be in a HSM or something, I don't know. The 'developer made a mistake' threat seems more real.

### PR DESCRIPTION
**Pull request by selfissued**
Fixes #105

### CODE REVIEW

**Comment by bc-pi** [comment]
one minor nit but otherwise looks good

File: main.md [hunk 1]
```diff
@@ -564,6 +564,17 @@ JWS `alg` values the authorization server supports for DPoP proof JWTs.
...
+## Client Registration Metadata {#client-meta}
+
+This document introduces the following new client registration metadata
+[@RFC7591] parameter to indicate that the client always uses
+DPoP when communicating with authorization servers.
+
+`always_uses_dpop`
+:   Boolean value specifying whether the client always uses DPoP.  If omitted, the default value is `false`.
+
+If `true`, the authorization server MUST reject requests from this client that do not contain the DPoP header.
```
[the remark this answers is not in the document]
Comment (danielfett) [reply main.md:576]: Does this imply that every AS with DPoP support must also know the client metadata of all its clients?

File: main.md [hunk 2]
```diff
@@ -564,6 +564,17 @@ JWS `alg` values the authorization server supports for DPoP proof JWTs.
 :   A JSON array containing a list of the JWS `alg` values supported
 by the authorization server for DPoP proof JWTs. 
 
+## Client Registration Metadata {#client-meta}
+
+This document introduces the following new client registration metadata
+[@RFC7591] parameter to indicate that the client always uses
+DPoP when communicating with authorization servers.
```
Comment (bc-pi) [inline main.md:571]:
```suggestion
DPoP when communicating the authorization server.
```

Comment (bc-pi) [reply main.md:571]: or ... 
```suggestion
DPoP when requesting tokens from the authorization server.
```

File: main.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (bc-pi) [reply main.md:576]: I'm not sure I understand the question... Client metadata is information that the AS has about the client. This is just one more bit of metadata.

**Review by b---c — DISMISSED** [review]
LGTM with the 2nd commit

File: main.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielfett) [reply main.md]: The AS not always has a full view on all the metadata the client has. This clause can be construed as requiring an AS to always ensure it has all metadata about the client there is (e.g., by requiring dynamic client registration).

**Comment by panva** [comment]
> Boolean value specifying whether the client always uses DPoP for token requests.

As is `always_uses_dpop` does not force [PAR requests](#111) from the client to include the `DPoP` header.

I realize the language states `token requests` so my question is, is that intentionally leaving out the PAR interactions introduced in #111, or was it not considered at the time of writing?

Either is fine with me but I would expect DPoP use to be enforced for PAR requests from for `always_uses_dpop` clients so that, in doing so, the authorization code is always bound to the private key.

File: main.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (bc-pi) [reply main.md]: Would wording similar to PAR's `require_pushed_authorization_requests` https://www.rfc-editor.org/rfc/rfc9126.html#name-client-metadata help qualify?

**Comment by danielfett** [comment]
> > Boolean value specifying whether the client always uses DPoP for token requests.
> 
> As is `always_uses_dpop` does not force [PAR requests](#111) from the client to include the `DPoP` header.
> 
> I realize the language states `token requests` so my question is, is that intentionally leaving out the PAR interactions introduced in #111, or was it not considered at the time of writing?
> 
> Either is fine with me but I would expect DPoP use to be enforced for PAR requests from for `always_uses_dpop` clients so that, in doing so, the authorization code is always bound to the private key.

I would expect that `always_uses_dpop` means that either `dpop_jkt` or the `DPoP` header need to be present in the PAR request.

**Comment by bc-pi** [comment]
> I realize the language states token requests so my question is, is that intentionally leaving out the PAR interactions introduced in https://github.com/danielfett/draft-dpop/pull/111, or was it not considered at the time of writing?

It was written before the PAR text but also kinda intentional in that it was meant to convey that only dpop bound ATs and when applicable RTs would be issued for the given client. But perhaps that's too narrow... ? 

> I would expect that always_uses_dpop means that either dpop_jkt or the DPoP header need to be present in the PAR request.

That's not an unreasonable expectation. But not what the text currently says. 

If always_uses_dpop was expanded/reworded to mean that, it seems like it should also mean that dpop_jkt needs to be present in regular authz request. ?

**Comment by b---c** [comment]
Trying to pull up this ongoing conversion in a resolved thread so it's not lost: 

> [danielfett](https://github.com/danielfett) [10 hours ago](https://github.com/danielfett/draft-dpop/pull/109#discussion_r808878902)
The AS not always has a full view on all the metadata the client has. This clause can be construed as requiring an AS to always ensure it has all metadata about the client there is (e.g., by requiring dynamic client registration).

 
> @[bc-pi](https://github.com/bc-pi) bc-pi [5 hours ago](https://github.com/danielfett/draft-dpop/pull/109#discussion_r809200182)
Would wording similar to PAR's require_pushed_authorization_requests https://www.rfc-editor.org/rfc/rfc9126.html#name-client-metadata help qualify?

**Comment by danielfett** [comment]
> Trying to pull up this ongoing conversion in a resolved thread so it's not lost:
> 
> > [danielfett](https://github.com/danielfett) [10 hours ago](https://github.com/danielfett/draft-dpop/pull/109#discussion_r808878902)
> > The AS not always has a full view on all the metadata the client has. This clause can be construed as requiring an AS to always ensure it has all metadata about the client there is (e.g., by requiring dynamic client registration).
> 
> > @[bc-pi](https://github.com/bc-pi) bc-pi [5 hours ago](https://github.com/danielfett/draft-dpop/pull/109#discussion_r809200182)
> > Would wording similar to PAR's require_pushed_authorization_requests https://www.rfc-editor.org/rfc/rfc9126.html#name-client-metadata help qualify?

Yes, that would be sufficient.

**Comment by danielfett** [comment]
> That's not an unreasonable expectation. But not what the text currently says.
> 
> If always_uses_dpop was expanded/reworded to mean that, it seems like it should also mean that dpop_jkt needs to be present in regular authz request. ?

That would be counter-intuitive, indeed. Thinking about this a bit more, keeping `always_uses_dpop` to mean that "the access token MUST be bound using DPoP" seems more logical, even though it might be a little surprising that there can be PAR requests without `dpop_jk` and `DPoP` header. But I would be fine with that.

File: main.md [hunk 3]
```diff
@@ -564,6 +564,17 @@ JWS `alg` values the authorization server supports for DPoP proof JWTs.
...
+
+This document introduces the following new client registration metadata
+[@RFC7591] parameter to indicate that the client always uses
+DPoP when requesting tokens from the authorization server.
+
+`always_uses_dpop`
+:   Boolean value specifying whether the client always uses DPoP for token requests.  If omitted, the default value is `false`.
+
+If `true`, the authorization server MUST reject token requests from this client that do not contain the DPoP header.
+
```
Comment (b---c) [inline main.md:584]:
```suggestion
## Client Metadata {#client-meta}

The Dynamic Client Registration Protocol [@RFC7591] defines an API for dynamically registering OAuth 2.0 client metadata with authorization servers. The metadata defined by [@RFC7591], and registered extensions to it, also imply a general data model for clients that is useful for authorization server implementations even when the Dynamic Client Registration Protocol isn't in play. Such implementations will typically have some sort of user interface available for managing client configuration. The following client metadata parameter is introduced by this document to indicate that the client always uses DPoP when requesting tokens from the authorization server.

`always_uses_dpop`
:   Boolean value specifying whether the client always uses DPoP for token requests.  If omitted, the default value is `false`.

If `true`, the authorization server MUST reject token requests from this client that do not contain the DPoP header.
```

Comment (b---c) [reply main.md:584]: The suggestion above is meant to replace the whole subsection. The change is based on a and meant to address [this piecemeal conversation](https://github.com/danielfett/draft-dpop/pull/109#issuecomment-1043406960) with @danielfett

Comment (selfissued) [reply main.md:584]: The third commit https://github.com/danielfett/draft-dpop/pull/109/commits/63fab689a42fc79b77ebdf98a73b41f7b4d69ab7, which I apparently made in parallel with your suggestion above, does exactly what you requested, modulo different line breaks.

Are we good to merge?

Comment (bc-pi) [reply main.md:584]: Damn, that was confusing... but yeah, I think we are we good to merge this one.

Comment (selfissued) [reply main.md:584]: Can I get an approval?

Comment (selfissued) [reply main.md:584]: Yep - I understand we may tweak the meaning but I think doing that with `always_use_dpop` and the PAR text in the same draft will be easier.

**Review by b---c — APPROVED** [review]

**Pull request merged**